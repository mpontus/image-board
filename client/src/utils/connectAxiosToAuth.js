import { Observable, empty } from "rxjs";
import { switchMap } from "rxjs/operators";
import { set, lensPath } from "ramda";

/**
 * Connects axios instance AuthService using interceptors
 *
 * @param axios Axios
 * @param authService AuthService
 * @return fn Unsubscribe function
 */
const connectAxiosToAuth = (axios, authService) =>
  authService
    .getIdToken()
    .pipe(
      switchMap(
        token =>
          token
            ? Observable.create(() => {
                const interceptor = set(
                  lensPath(["headers", "Authorization"]),
                  `Bearer ${token}`
                );

                axios.interceptors.request.use(interceptor);

                return () => axios.interceptors.request.eject(interceptor);
              })
            : empty()
      )
    )
    .subscribe();
export default connectAxiosToAuth;
