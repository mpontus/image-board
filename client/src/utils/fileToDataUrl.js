/**
 * Convert Blob or File to data url containing Base64 string.
 *
 * @param file File | Blob Blob or file
 * @return Promise<string> Data URL
 */
const fileToDataUrl = file =>
  new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => resolve(reader.result);
    reader.onerror = reject;
  });

export default fileToDataUrl;
