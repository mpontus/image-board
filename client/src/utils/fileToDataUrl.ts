/**
 * Convert File to data url
 *
 * @param file File | Blob Blob or file
 * @return Promise<string> Data URL
 */
const fileToDataUrl = (file: File): Promise<string> =>
  new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => {
      if (reader.result == null) {
        reject(new Error("Reader result is null"));
      } else if (reader.result instanceof ArrayBuffer) {
        resolve(reader.result.toString());
      } else {
        resolve(reader.result);
      }
    };
    reader.onerror = reject;
  });

export default fileToDataUrl;
