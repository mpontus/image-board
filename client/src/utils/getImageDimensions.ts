interface Dimensions {
  width: number;
  height: number;
}

/**
 * Return dimensions of an image specified as a File
 *
 * @param file File Image file
 * @return Dimensions Object containing width and height
 */
const getImageDimensions = (file: File): Promise<Dimensions> =>
  new Promise((resolve, reject) => {
    const image = new Image();
    image.src = URL.createObjectURL(file);
    image.onload = () =>
      resolve({
        width: image.width,
        height: image.height,
      });
  });

export default getImageDimensions;
