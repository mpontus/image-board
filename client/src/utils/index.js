export { default as fileToDataUrl } from "./fileToDataUrl";
export { default as connectAxiosToAuth } from "./connectAxiosToAuth";
