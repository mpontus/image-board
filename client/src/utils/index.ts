export { default as getImageDimensions } from "./getImageDimensions";
export { default as fileToDataUrl } from "./fileToDataUrl";
