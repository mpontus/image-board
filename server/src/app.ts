import * as cors from "cors";
import * as dotenv from "dotenv-safe";
import * as express from "express";
import * as jwt from "express-jwt";
import * as paginate from "express-paginate";
import * as jwksRsa from "jwks-rsa";
import * as mongoose from "mongoose";
import * as morgan from "morgan";
import * as multer from "multer";
import * as shortid from "shortid";
import * as winston from "winston";
import Post from "./model/Post";
import createStorage from "./storage/createStorage";

dotenv.config();

const storage = createStorage({
  storageType: process.env.CLOUDINARY_STUB ? "test" : "cloudinary",
  cloudinaryUrl: process.env.CLOUDINARY_URL,
  storageParams: {
    folder: process.env.CLOUDINARY_FOLDER || "",
    allowedFormats: ["jpg", "png"],
    filename: (req, file, cb) => cb(null, shortid())
  }
});

mongoose.connect(
  process.env.MONGODB_URI || "",
  {
    useNewUrlParser: true
  }
);

const upload = multer({
  storage
});

const logger = winston.createLogger({
  transports: [new winston.transports.Console()]
});

const app = express();

app.use(morgan("combined"));
app.use(cors());

app.use(
  jwt(
    process.env.NODE_ENV === "test"
      ? {
          secret: process.env.JWT_SECRET || "",
          credentialsRequired: false
        }
      : {
          secret: jwksRsa.expressJwtSecret({
            cache: true,
            rateLimit: true,
            jwksRequestsPerMinute: 5,
            jwksUri: `https://${process.env.AUTH0_DOMAIN}/.well-known/jwks.json`
          }),
          audience: process.env.AUTH0_CLIENT_ID,
          issuer: `https://${process.env.AUTH0_DOMAIN}/`,
          algorithms: ["RS256"],
          credentialsRequired: false
        }
  )
);

const requireAuth: express.RequestHandler = (req, res, next) => {
  if (!req.user) {
    res.status(401).end();
  } else {
    next();
  }
};

interface User {
  sub: string;
  nickname: string;
  picture: string;
}

type ModelType<T> = T extends mongoose.Model<infer R> ? R : any;

const serializePost = (user: User) => (post: ModelType<typeof Post>) => ({
  id: post._id,
  imageUrl: post.imageUrl,
  imageWidth: post.imageWidth,
  imageHeight: post.imageHeight,
  author: {
    id: post.author.id,
    name: post.author.name,
    avatarUrl: post.author.avatarUrl
  },
  likes: post.likesCount,
  isLiked: user ? post.likes.includes(user.sub) : false,
  timestamp: post.timestamp
});

app.get("/api/posts", paginate.middleware(5, 5), async (req, res, next) => {
  try {
    const [total, posts] = await Promise.all([
      Post.find({})
        .countDocuments()
        .exec(),

      Post.find({})
        .sort("-timestamp")
        .skip(req.skip as number)
        .limit(req.query.limit)
        .lean()
        .exec()
    ]);

    res.json({
      total,
      items: posts.map(serializePost(req.user))
    });
  } catch (error) {
    next(error);
  }
});

app.post(
  "/api/posts",
  requireAuth,
  upload.single("file"),
  async (req, res, next) => {
    try {
      if (!req.user.nickname || !req.user.picture) {
        res.status(400).send();
      }

      const file = req.file as any;

      const post = await Post.create({
        imageId: file.public_id,
        imageUrl: file.url,
        imageWidth: file.width,
        imageHeight: file.height,
        author: {
          id: req.user.sub,
          name: req.user.nickname,
          avatarUrl: req.user.picture
        },
        likes: [req.user.sub],
        likesCount: 1
      });

      res.status(201).json(serializePost(req.user)(post));
    } catch (error) {
      next(error);
    }
  }
);

app.delete("/api/posts/:id", requireAuth, async (req, res, next) => {
  try {
    const post = mongoose.Types.ObjectId.isValid(req.params.id)
      ? await Post.findOne({ _id: req.params.id })
      : null;

    if (post === null) {
      res.status(404).send();

      return;
    }

    if (req.user.sub !== post.author.id) {
      res.status(401).send();
    }

    await Promise.all([storage.delete(post.imageId), post.remove()]);

    res.status(204).end();
  } catch (error) {
    next(error);
  }
});

app.put("/api/posts/:id/like", requireAuth, async (req, res, next) => {
  try {
    const { nModified } = await Post.updateOne(
      {
        _id: req.params.id,
        likes: { $ne: req.user.sub }
      },
      {
        $inc: {
          likesCount: 1
        },
        $push: {
          likes: req.user.sub
        }
      }
    ).exec();

    if (nModified === 0) {
      res.status(400).end();

      return;
    }

    res.status(202).end();
  } catch (error) {
    next(error);
  }
});

app.delete("/api/posts/:id/like", requireAuth, async (req, res, next) => {
  try {
    const { nModified } = await Post.updateOne(
      {
        _id: req.params.id,
        likes: req.user.sub
      },
      {
        $inc: {
          likesCount: -1
        },
        $pull: {
          likes: req.user.sub
        }
      }
    ).exec();

    if (nModified === 0) {
      res.status(400).end();

      return;
    }

    res.status(202).end();
  } catch (error) {
    next(error);
  }
});

app.use(((err, req, res, next) => {
  logger.error(err);

  res.status(500).json({
    message: err.message
  });
}) as express.ErrorRequestHandler);

export default app;
