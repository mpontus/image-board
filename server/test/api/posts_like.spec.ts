/* tslint:disable:no-shadowed-variable */
import app from "@src/app";
import * as request from "supertest";
import { createToken, resetDb } from "../utils";

beforeEach(() => resetDb());

describe("/api/posts/:id/like", () => {
  describe("PUT", () => {
    const {
      postId,
      authorId,
      default: seed
    } = require("@test/seed/single-post");

    describe("when post exists", () => {
      beforeEach(seed);

      it("should return 202 Accepted", async () => {
        const response = await request(app)
          .put(`/api/posts/${postId}/like`)
          .set("Authorization", `Bearer ${createToken({ sub: authorId })}`)
          .expect(202);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe("when post doesn't exist", () => {
      it("should return 400 Bad Request", async () => {
        const response = await request(app)
          .put(`/api/posts/${postId}/like`)
          .set("Authorization", `Bearer ${createToken({ sub: authorId })}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe("when user is unauthenticated", () => {
      beforeEach(seed);

      it("should return 401 Unauthorized", async () => {
        const response = await request(app)
          .put(`/api/posts/${postId}/like`)
          .expect(401);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe("when post is already liked", () => {
      const {
        postId,
        authorId,
        default: seed
      } = require("@test/seed/liked-post");

      beforeEach(seed);

      it("should return 400 Bad Request", async () => {
        const response = await request(app)
          .put(`/api/posts/${postId}/like`)
          .set("Authorization", `Bearer ${createToken({ sub: authorId })}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe("DELETE", () => {
    const {
      postId,
      authorId,
      default: seed
    } = require("@test/seed/liked-post");

    describe("when post exists", () => {
      beforeEach(seed);

      it("should return 202 Accepted", async () => {
        const response = await request(app)
          .delete(`/api/posts/${postId}/like`)
          .set("Authorization", `Bearer ${createToken({ sub: authorId })}`)
          .expect(202);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe("when post doesn't exist", () => {
      it("should return 400 Bad Request", async () => {
        const response = await request(app)
          .delete(`/api/posts/${postId}/like`)
          .set("Authorization", `Bearer ${createToken({ sub: authorId })}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe("when user is unauthenticated", () => {
      beforeEach(seed);

      it("should return 401 Unauthorized", async () => {
        const response = await request(app)
          .delete(`/api/posts/${postId}/like`)
          .expect(401);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe("when post has not been liked", () => {
      const {
        postId,
        authorId,
        default: seed
      } = require("@test/seed/single-post");

      beforeEach(seed);

      it("should return 400 Bad Request", async () => {
        const response = await request(app)
          .delete(`/api/posts/${postId}/like`)
          .set("Authorization", `Bearer ${createToken({ sub: authorId })}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });
  });
});
