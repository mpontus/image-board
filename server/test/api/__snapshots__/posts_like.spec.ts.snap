// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/posts DELETE when post doesn't exist should return 400 Bad Request 1`] = `Object {}`;

exports[`/api/posts DELETE when post exists should return 202 Accepted 1`] = `Object {}`;

exports[`/api/posts DELETE when post has not been liked should return 400 Bad Request 1`] = `Object {}`;

exports[`/api/posts DELETE when user is unauthenticated should return 401 Unauthorized 1`] = `Object {}`;

exports[`/api/posts PUT when post doesn't exist should return 400 Bad Request 1`] = `Object {}`;

exports[`/api/posts PUT when post exists should return 202 Accepted 1`] = `Object {}`;

exports[`/api/posts PUT when post is already liked should return 400 Bad Request 1`] = `Object {}`;

exports[`/api/posts PUT when user is unauthenticated should return 401 Unauthorized 1`] = `Object {}`;
